site/source/docs/porting/exceptions.rst

```diff
@@ -32,19 +23,54 @@ For example, build the following program with ``emcc -fexceptions``:
       return 0;
     }
 
-Running it will print ``throw...`` and then ``catch!``.
+The first ``throw`` will abort the program and you'll see something like this in the output:
+
+.. code-block:: text
+
+  throw...  
+  exception thrown: 5246024 - Exception catching is disabled, this exception cannot be caught. Compile with -s NO_DISABLE_EXCEPTION_CATCHING or -s EXCEPTION_CATCHING_ALLOWED=[..] to catch.
+
+If you want to opt-in, you have two following options.
+
+JavaScript-based exception support
+##################################
+
+First, you can enable exceptions via Emscripten's JavaScript-based support.
+
+To enable it, pass ``-fexceptions`` at both compile time and link time.
+
+When you rebuild the example above with this flag, the output will change to:
+
+.. code-block:: text
+
+  throw...
+  catch!
+
+Note that this option has relatively high overhead, but it will work on all
+engines with WebAssembly support. You can reduce the overhead by specifying a
+list of allowed functions in which exceptions are enabled, see the
+``EXCEPTION_CATCHING_ALLOWED`` setting.
+
+WebAssembly exception handling proposal
+#######################################
+
+Alternatively, you can opt-in to the `native WebAssembly exception handling
+<https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md>`_.
+
+To enable it, pass ``-fwasm-exceptions`` at both compile time and link time.
+
+Rebuilding the example with this flag will result in the same output as with
+``-fexceptions`` above:
 
-Overhead
-########
+.. code-block:: text
 
-Until wasm gets direct support for exceptions, Emscripten supports them using
-JavaScript for throwing and catching. As a result, every time we might need to
-catch an exception we call out into JavaScript to a try-catch there. This
-increases code size and adds significant overhead.
+  throw...
+  catch!
 
-You can reduce the overhead by specifying a list of allowed functions in
-which exceptions are enabled, see the ``EXCEPTION_CATCHING_ALLOWED`` setting.
+This option leverages a new feature that brings built-in instructions for
```

Comment:
Maybe say "a new WebAssembly feature" to distinguish between toolchain and wasm features

Reply:
It seems a bit repetitive because I do say "brings built-in instructions [...] to WebAssembly". Maybe there's a better rewording?

Reply:
Actually re-reading the whole thing in context, the original isn't so bad.
So that would be ok; or we could perhaps mention the engine (rather than "WebAssembly" abstractly) to disambiguate, e.g. something like "leverages a new WebAssembly feature that brings built-in instructions for throwing and catching exceptions directly to the engine"

Reply:
It feels a bit confusing, because instructions are added to the WebAssembly the spec rather than any specific engine. Maybe I should drop that part and reword to just "leverages a new WebAssembly feature that adds built-in instructions for throwing and catching exceptions"?

Reply:
I'd rather keep this bit with motivation from the last comment.